### PR TITLE
Move plugin-related utilities and API to SDK/tracer

### PR DESCRIPTION
**Pull request by raphw**
This API separates API that is mainly used within plugins from the *core* module but makes it available via the *tracer-api* (if functional) or the *sdk* module (if technical). The scope of the latter can be discussed as this exposes it to user plugins, but I found that there are several legitimate uses of some of the Byte Buddy extensions (several custom annotations and matchers) from within non-built-in-plugins. This is also true for several *util* classes which are mainly used from plugins where technical utilities are moved to the SDK and functional utilities are moved to the tracer module. Also, the `PluginClassLoaderRootPackageCustomizer` is moved to the SDK (the question being if it should be exposed, but I think there are use cases). Finally, the `TraceAwareInstrumentation` class is removed as it creates an unfortunate binding between the SDK and tracer module which does not otherwise exist elsewhere. By removing this class, the two modules can remain separated.

### CODE REVIEW

**Comment by raphw** [comment]
I resolved the import conflicts. @SylvainJuge - did you have time to review? Almost there!

**Comment by SylvainJuge** [comment]
Hi @raphw , thanks for keeping this PR up-to-date.
I've already had a couple of large PR reviews this week, but I'll try to prioritize yours next week.

**Review by SylvainJuge — APPROVED** [review]
Yet another great refactor, thanks for this @raphw !

I think it's mostly fine to merge, and I only have very minor comments.

The extra indirection with service providers seems a bit overkill, but I don't think we have a better way to implement that without exposing implementation details from the core module to the plugins sdk, so I think it's fine. As it's only a few static methods maybe having a single provider/implementation could maybe simplify it a bit, but that would just group unrelated methods together in a single interface.

**Comment by JonasKunz** [comment]
LGTM as well, I just noticed that a lot of stuff has been moved into the sdk-project, which is intended to be used as public api for external plugins.
Should we maybe move those classes to the `.internal` package, to mark them as not so public?

And regarding the `PluginClassLoaderRootPackageCustomizer`: Even if it is now visible to external plugins, it won't be used. For external plugins we always copy all of the classes from the external plugin jar into the plugin classloaders.
For that reason, I would move that one into an `internal` package aswell.

**Comment by jackshirazi** [comment]
we shouldn't be expanding the public API at all, as it's intended to be deprecated

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bbwarmup/WarmupInstrumentation.java [hunk 1]
```diff
@@ -31,13 +32,17 @@
...
 import static net.bytebuddy.matcher.ElementMatchers.named;
 import static net.bytebuddy.matcher.ElementMatchers.not;
 import static net.bytebuddy.matcher.ElementMatchers.returns;
 import static net.bytebuddy.matcher.ElementMatchers.takesNoArguments;
 
-public class WarmupInstrumentation extends TracerAwareInstrumentation {
+public class WarmupInstrumentation extends ElasticApmInstrumentation {
+
+    @SuppressWarnings("unused")
+    private static final Tracer tracer = GlobalTracer.get();
```
[the remark this answers is not in the document]
Comment (raphw) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/bbwarmup/WarmupInstrumentation.java]: That's correct, I changed it.

**Comment by raphw** [comment]
The premise was that it should be possible to implement Elastic's plugins with the SDK and tracer APIs. As some of the APIs build on top of the current SDK, this was the only way to go. Another option would be to define a *sdk-ext* module which houses these extensions and to declare them not safe for public use. If one wanted to make these classes available later, they could still be moved to the *non-ext* module.

**Comment by JonasKunz** [comment]
> we shouldn't be expanding the public API at all, as it's intended to be deprecated in favour of the OpenTelemetry API

I think we need to distinguish between the API used for creating spans (=opentelemetry or our "public API") vs the API required for defining the bootstraping of the external plugin, e.g. `ElasticApmInstrumentation`. The latter one is represented by our `sdk` project and there is no compatible OTel pendant yet to my knowledge.

**Comment by JonasKunz** [comment]
> Another option would be to define a sdk-ext module which houses these extensions and to declare them not safe for public use. If one wanted to make these classes available later, they could still be moved to the non-ext module.

I think that would be a bit overkill and add yet another module, of which we already have plenty 😅 . Therefore my suggestion to mark them as not intended for public uses by moving them to an `internal` package within the same module. While this does not prevent usage in external plugins, it at least discourages it.

**Comment by raphw** [comment]
In Mockito, we use an [`Incubating` annotation](https://javadoc.io/static/org.mockito/mockito-core/2.3.7/org/mockito/Incubating.html) which is more flexible as it can be placed on methods, too. But I am happy to adjust either way.

**Comment by JonasKunz** [comment]
run elasticsearch-ci/docs

**Pull request merged**